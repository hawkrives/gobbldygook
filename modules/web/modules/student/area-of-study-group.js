import React, { PropTypes } from 'react'
import map from 'lodash/map'
import pluralizeArea from '../../../examine-student/pluralize-area'
import capitalize from 'lodash/capitalize'
import { areaTypeConstants } from '../../../object-student/area-types'
import values from 'lodash/values'

import AreaOfStudy from '../area-of-study'
import AreaPicker from './area-picker'
import Button from '../../components/button'

import './area-of-study-group.scss'

export default function AreaOfStudyGroup(props) {
    const showAreaPicker = props.showAreaPicker || false
    const showOrHidePicker = showAreaPicker
        ? props.onEndAddArea
        : props.onInitiateAddArea

    return (
        <section className="area-of-study-group">
            <h1 className="area-type-heading">
                {capitalize(pluralizeArea(props.type))}
                <Button
                    className="add-area-of-study"
                    type="flat"
                    onClick={ev => showOrHidePicker(props.type, ev)}
                >
                    {showAreaPicker ? 'Close' : 'Add âˆ™ Edit'}
                </Button>
            </h1>

            {showAreaPicker
                ? <AreaPicker
                      areaList={props.allAreasOfType}
                      currentAreas={props.areas}
                      onAddArea={props.onAddArea}
                      studentGraduation={props.studentGraduation}
                      type={props.type}
                  />
                : null}

            {map(props.areas, (area, i) => (
                <AreaOfStudy
                    area={area}
                    key={i + area.name ? area.name : ''}
                    onAddOverride={props.onAddOverride}
                    onRemoveArea={props.onRemoveArea}
                    onRemoveOverride={props.onRemoveOverride}
                    onToggleOverride={props.onToggleOverride}
                    showCloseButton={showAreaPicker}
                    showEditButton={showAreaPicker}
                    studentId={props.studentId}
                />
            ))}
        </section>
    )
}

AreaOfStudyGroup.propTypes = {
    allAreasOfType: PropTypes.arrayOf(PropTypes.object).isRequired,
    areas: PropTypes.arrayOf(PropTypes.object).isRequired,
    onAddArea: PropTypes.func.isRequired,
    onAddOverride: PropTypes.func.isRequired,
    onEndAddArea: PropTypes.func.isRequired,
    onInitiateAddArea: PropTypes.func.isRequired,
    onRemoveArea: PropTypes.func.isRequired,
    onRemoveOverride: PropTypes.func,
    onToggleOverride: PropTypes.func.isRequired,
    showAreaPicker: PropTypes.bool.isRequired,
    studentGraduation: PropTypes.number.isRequired,
    studentId: PropTypes.string.isRequired,
    type: PropTypes.oneOf(values(areaTypeConstants)).isRequired,
}
