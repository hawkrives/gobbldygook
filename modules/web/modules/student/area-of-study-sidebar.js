// @flow
import React from 'react'
import { connect } from 'react-redux'

import difference from 'lodash/difference'
import filter from 'lodash/filter'
import find from 'lodash/find'
import groupBy from 'lodash/groupBy'
import includes from 'lodash/includes'
import keys from 'lodash/keys'
import map from 'lodash/map'
import mapValues from 'lodash/mapValues'
import pick from 'lodash/pick'
import pickBy from 'lodash/pickBy'
import toPairs from 'lodash/toPairs'
import union from 'lodash/union'
import uniq from 'lodash/uniq'
import values from 'lodash/values'

import AreaOfStudyGroup from './area-of-study-group'
import Button from '../../components/button'
const {
    sortStudiesByType,
} = require('../../../object-student/sort-studies-by-type')
const { areaTypeConstants } = require('../../../object-student/area-types')

type AreaOfStudy = Object
type Student = Object
type PropTypes = {
    allAreas: AreaOfStudy[],
    student: Student,
}

class AreaOfStudySidebarComponent extends React.PureComponent {
    props: PropTypes

    state: {
        showAreaPickerFor: { [key: string]: boolean },
    } = {
        showAreaPickerFor: {},
    }

    showAreaPicker = (type: string, ev: Event) => {
        ev.stopPropagation()
        ev.preventDefault()
        this.setState(state => ({
            showAreaPickerFor: { ...state.showAreaPickerFor, [type]: true },
        }))
    }

    hideAreaPicker = (type: string, ev: Event) => {
        ev.stopPropagation()
        ev.preventDefault()
        this.setState(state => ({
            showAreaPickerFor: { ...state.showAreaPickerFor, [type]: false },
        }))
    }

    render() {
        const props = this.props
        const { allAreas, student } = props
        const { showAreaPickerFor } = this.state
        const allAreasGrouped = groupBy(allAreas, 'type')

        const sortedStudies = sortStudiesByType(student.studies)

        // group the studies by their type
        const groupedStudies = groupBy(sortedStudies, study =>
            study.type.toLowerCase()
        )

        // pull out the results
        const studyResults = mapValues(groupedStudies, group =>
            map(
                group,
                area =>
                    find(
                        student.areas,
                        pick(area, ['name', 'type', 'revision'])
                    ) || area
            )
        )

        // and then render them
        const sections = map(studyResults, (areas, areaType) => (
            <AreaOfStudyGroup
                key={areaType}
                allAreasOfType={allAreasGrouped[areaType] || []}
                areas={areas}
                onEndAddArea={this.hideAreaPicker}
                onInitiateAddArea={this.showAreaPicker}
                showAreaPicker={showAreaPickerFor[areaType] || false}
                student={student}
                type={areaType}
            />
        ))

        const allAreaTypes = values(areaTypeConstants)
        const usedAreaTypes = uniq(map(student.studies, s => s.type))

        const areaTypesToShowButtonsFor = union(
            usedAreaTypes,
            keys(pickBy(showAreaPickerFor, v => v === true))
        )

        const unusedTypes = difference(allAreaTypes, areaTypesToShowButtonsFor)

        const unusedAreaTypeButtons = unusedTypes.length
            ? <section className="unused-areas-of-study">
                  <span className="unused-areas-title">Add: </span>
                  <span className="unused-areas-buttons">
                      {unusedTypes.map(type => (
                          <Button
                              key={type}
                              className="add-unused-area-of-study"
                              onClick={ev => this.showAreaPicker(type, ev)}
                              type="flat"
                          >
                              {type}
                          </Button>
                      ))}
                  </span>
              </section>
            : null

        const unusedTypesToShow = filter(
            toPairs(showAreaPickerFor),
            ([type, toShow]) =>
                toShow === true && !includes(usedAreaTypes, type)
        )

        const unusedTypesToShowComponents = map(unusedTypesToShow, ([
            type,
            shouldShow,
        ]) => (
            <AreaOfStudyGroup
                key={type}
                allAreasOfType={allAreasGrouped[type] || []}
                areas={[]}
                onEndAddArea={this.hideAreaPicker}
                onInitiateAddArea={this.showAreaPicker}
                showAreaPicker={shouldShow || false}
                student={student}
                type={type}
            />
        ))

        return (
            <div>
                {sections}
                {unusedTypesToShowComponents}
                {unusedAreaTypeButtons}
            </div>
        )
    }
}

const mapState = state => ({ allAreas: state.areas.data })

export const AreaOfStudySidebar = connect(mapState)(AreaOfStudySidebarComponent)
