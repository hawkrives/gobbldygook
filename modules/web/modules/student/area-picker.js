import React, { Component, PropTypes } from 'react'
import fuzzysearch from 'fuzzysearch'
import pluralizeArea from '../../../examine-student/pluralize-area'
import map from 'lodash/map'
import reject from 'lodash/reject'
import filter from 'lodash/filter'
import includes from 'lodash/includes'
import { filterAreaList } from '../../../object-student/filter-area-list'

import Button from '../../components/button'
import List from '../../components/list'
import Toolbar from '../../components/toolbar'

import './area-picker.scss'

function AreaPicker(props) {
    const graduation = props.studentGraduation

    const currentAreaNames = map(props.currentAreas, a => a.name)
    let onlyAvailableAreas = reject(props.areaList, area =>
        includes(currentAreaNames, area.name))

    onlyAvailableAreas = filterAreaList(onlyAvailableAreas, { graduation })

    const filteredOnName = filter(onlyAvailableAreas, area =>
        fuzzysearch(props.filterText, area.name.toLowerCase()))

    const areaList = map(filteredOnName, (area, i) => (
        <li key={area.name + i} className="area--choice">
            <span className="area-listing">
                <span className="title">{area.name}</span>
                <span className="revision">{area.revision}</span>
            </span>
            <Button
                className="toggle-area"
                type="flat"
                onClick={ev => props.onAddArea(area, ev)}
            >
                Add
            </Button>
        </li>
    ))

    let message = <li>Oh! We need a new message here!</li>
    if (props.filterText) {
        message = <li>No matching ${pluralizeArea(props.type)}.</li>
    } else if (currentAreaNames.size) {
        message = <li>All ${pluralizeArea(props.type)} have been added.</li>
    } else {
        message = <li>No {pluralizeArea(props.type)} are available.</li>
    }

    return (
        <div className="add-area">
            <Toolbar>
                <input
                    className="add-area--filter"
                    placeholder={`Filter ${pluralizeArea(props.type)}`}
                    value={props.filterText}
                    onChange={props.onFilterChange}
                />
            </Toolbar>

            <List type="plain">
                {areaList.length ? areaList : message}
            </List>
        </div>
    )
}

AreaPicker.propTypes = {
    areaList: PropTypes.arrayOf(PropTypes.object).isRequired,
    currentAreas: PropTypes.arrayOf(PropTypes.object).isRequired,
    filterText: PropTypes.string.isRequired,
    onAddArea: PropTypes.func.isRequired,
    onFilterChange: PropTypes.func.isRequired,
    studentGraduation: PropTypes.number.isRequired,
    type: PropTypes.string.isRequired,
}

export default class AreaPickerContainer extends Component {
    state = {
        filter: '',
    };

    render() {
        return (
            <AreaPicker
                {...this.props}
                filterText={this.state.filter}
                onFilterChange={ev =>
                    this.setState({
                        filter: (ev.target.value || '').toLowerCase(),
                    })}
            />
        )
    }
}
