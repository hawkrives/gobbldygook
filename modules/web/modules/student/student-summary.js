// @flow
import React from 'react'
import cx from 'classnames'
import listify from 'listify'
import groupBy from 'lodash/groupBy'
import sample from 'lodash/sample'

import AvatarLetter from '../../components/avatar-letter'
import ContentEditable from '../../components/content-editable'

import {
    getActiveStudentCourses,
} from '../../helpers/get-active-student-courses'
import { countCredits } from '../../../examine-student/count-credits'

import './student-summary.scss'

const welcomeMessages = [
    'Hi, ',
    'Hi there, ',
    'Hello, ',
    'こんにちは、', // japanese
    'ようこそ、', // japanese
    'Fram! Fram! ',
    'Salut, ',
    'Aloha, ',
    'Привет, ',
    'Вітаю, ',
    'Sawubona, ',
    'Hei, ',
    'Hola, ', // spanish
    'Bonjour, ', // french
    'Hallo, ', // german
    'nyob zoo ', // hmong
    '你好，', // mandarin
    '안녕하세요 ', // korean
    'สวัสดี ', // thai
    'halo, ', // indonesian
]

const welcomeMessage = welcomeMessages[2]

type Student = Object;

export class StudentSummary extends React.PureComponent {
    props: {
        onChangeGraduation?: string => any,
        onChangeMatriculation?: string => any,
        onChangeName?: string => any,
        randomizeHello?: boolean,
        showAvatar?: boolean,
        showMessage?: boolean,
        student: Student,
    };

    render() {
        const {
            student,
            showMessage = true,
            showAvatar = true,
            randomizeHello = true,
            ...props
        } = this.props
        const { studies, canGraduate } = student

        const className = canGraduate ? 'can-graduate' : 'cannot-graduate'

        const currentCredits = countCredits(getActiveStudentCourses(student))
        const neededCredits = student.creditsNeeded

        const message = randomizeHello
            ? sample(welcomeMessages)
            : welcomeMessage

        return (
            <article className={cx('student-summary', className)}>
                <Header
                    canGraduate={canGraduate}
                    name={student.name}
                    onChangeName={props.onChangeName}
                    helloMessage={message}
                    showAvatar={showAvatar}
                />
                <div className="content">
                    <DateSummary
                        onChangeGraduation={props.onChangeGraduation}
                        onChangeMatriculation={props.onChangeMatriculation}
                        matriculation={student.matriculation}
                        graduation={student.graduation}
                    />

                    <DegreeSummary studies={studies} />

                    <CreditSummary
                        currentCredits={currentCredits}
                        neededCredits={neededCredits}
                    />

                    {showMessage ? <Footer canGraduate={canGraduate} /> : null}
                </div>
            </article>
        )
    }
}

export class Header extends React.PureComponent {
    props: {
        canGraduate: boolean,
        helloMessage: string,
        name: string,
        onChangeName?: string => any,
        showAvatar: boolean,
    };

    render() {
        const props = this.props

        const className = props.canGraduate
            ? 'can-graduate'
            : 'cannot-graduate'

        const avatar = props.showAvatar
            ? <AvatarLetter
                  className={cx('student-letter', className)}
                  value={props.name}
              />
            : null

        const name = (
            <ContentEditable
                disabled={!props.onChangeName}
                className="autosize-input"
                onBlur={props.onChangeName}
                value={String(props.name)}
            />
        )

        return (
            <header className="student-summary--header">
                {avatar}

                <div className="intro">
                    {props.helloMessage}{name}!
                </div>
            </header>
        )
    }
}

export class Footer extends React.PureComponent {
    goodGraduationMessage = [
        "It looks like you'll make it! Just follow the plan, and",
        'go over my output with your advisor a few times.',
    ].join(' ');

    badGraduationMessage = [
        "You haven't planned everything out yet.",
        'Ask your advisor if you need help fitting everything in.',
    ].join(' ');

    props: {
        canGraduate: boolean,
    };

    render() {
        const msg = this.props.canGraduate
            ? this.goodGraduationMessage
            : this.badGraduationMessage

        return (
            <div className="paragraph graduation-message">
                {msg}
            </div>
        )
    }
}

export class DateSummary extends React.PureComponent {
    props: {
        onChangeGraduation?: string => any,
        onChangeMatriculation?: string => any,
        matriculation: string,
        graduation: string,
    };

    render() {
        const props = this.props

        const matriculation = (
            <ContentEditable
                disabled={!props.onChangeMatriculation}
                className="autosize-input"
                onBlur={props.onChangeMatriculation}
                value={String(props.matriculation)}
            />
        )

        const graduation = (
            <ContentEditable
                disabled={!props.onChangeGraduation}
                className="autosize-input"
                onBlur={props.onChangeGraduation}
                value={String(props.graduation)}
            />
        )

        return (
            <p className="paragraph">
                After matriculating in {matriculation}, you are
                planning to graduate in {graduation}.
            </p>
        )
    }
}

export class DegreeSummary extends React.PureComponent {
    props: {
        studies: { type: string, name: string }[],
    };

    render() {
        const grouped: {
            [key: string]: { type: string, name: string }[],
        } = groupBy(this.props.studies, s => s.type)

        const {
            degree: dS = [],
            major: mS = [],
            concentration: cS = [],
            emphasis: eS = [],
        } = grouped

        const dCount = dS.length
        const mCount = mS.length
        const cCount = cS.length
        const eCount = eS.length

        const dWord = dCount === 1 ? 'degree' : 'degrees'
        const mWord = mCount === 1 ? 'major' : 'majors'
        const cWord = cCount === 1 ? 'concentration' : 'concentrations'
        const eWord = eCount === 1 ? 'emphasis' : 'emphases'

        const dEmph = dCount === 1 ? 'a ' : ''
        const mEmph = mCount === 1 ? 'a ' : ''
        const cEmph = cCount === 1 ? 'a ' : ''
        const eEmph = eCount === 1 ? 'an ' : ''

        const dList = listify(dS.map(d => d.name))
        const mList = listify(mS.map(m => m.name))
        const cList = listify(cS.map(c => c.name))
        const eList = listify(eS.map(e => e.name))

        return (
            <p className="paragraph">
                You are planning on{' '}
                {dCount > 0 ? `${dEmph}${dList} ${dWord}` : `no ${dWord}`}
                {mCount || cCount || eCount
                    ? mCount && (cCount || eCount) ? ', ' : ' and '
                    : ''}
                {mCount ? `${mEmph}${mWord} in ${mList}` : ''}
                {mCount && cCount ? ', and ' : ''}
                {cCount ? `${cEmph}${cWord} in ${cList}` : ''}
                {(mCount || cCount) && eCount ? ', ' : ''}
                {eCount ? `not to mention ${eEmph}${eWord} in ${eList}` : ''}
                {'.'}
            </p>
        )
    }
}

export class CreditSummary extends React.PureComponent {
    props: {
        currentCredits: number,
        neededCredits: number,
    };

    render() {
        const { currentCredits, neededCredits } = this.props
        const enoughCredits = currentCredits >= neededCredits

        return (
            <p className="paragraph">
                You have currently planned for {currentCredits}{' '}
                of your {neededCredits} required credits.
                {enoughCredits ? ' Good job!' : ''}
            </p>
        )
    }
}
