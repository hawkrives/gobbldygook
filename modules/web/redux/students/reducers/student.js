import undoable from 'redux-undo'

import {
    changeStudentName,
    changeStudentAdvisor,
    changeStudentCreditsNeeded,
    changeStudentMatriculation,
    changeStudentGraduation,
    changeStudentSetting,
    addAreaToStudent,
    removeAreaFromStudent,
    addScheduleToStudent,
    destroyScheduleFromStudent,
    moveCourseToSchedule,
    setOverrideOnStudent,
    removeOverrideFromStudent,
    addFabricationToStudent,
    removeFabricationFromStudent,
    renameScheduleInStudent,
    reorderScheduleInStudent,
    moveScheduleInStudent,
    addCourseToSchedule,
    removeCourseFromSchedule,
    reorderCourseInSchedule,
} from '../../../../object-student/student'

import {
    LOAD_STUDENT,
    INIT_STUDENT,
    IMPORT_STUDENT,
    CHANGE_NAME,
    CHANGE_ADVISOR,
    CHANGE_CREDITS_NEEDED,
    CHANGE_MATRICULATION,
    CHANGE_GRADUATION,
    CHANGE_SETTING,
    ADD_AREA,
    REMOVE_AREA,
    REMOVE_AREAS,
    ADD_SCHEDULE,
    DESTROY_SCHEDULE,
    DESTROY_SCHEDULES,
    RENAME_SCHEDULE,
    REORDER_SCHEDULE,
    MOVE_SCHEDULE,
    ADD_COURSE,
    REMOVE_COURSE,
    REORDER_COURSE,
    MOVE_COURSE,
    SET_OVERRIDE,
    REMOVE_OVERRIDE,
    ADD_FABRICATION,
    REMOVE_FABRICATION,
} from '../constants'

const initialState = {}

export function studentReducer(state = initialState, action) {
    const { type, payload } = action

    switch (type) {
        case INIT_STUDENT: {
            if (action.error) {
                return state
            }
            return payload
        }

        case CHANGE_NAME: {
            return changeStudentName(state, payload.name)
        }
        case CHANGE_ADVISOR: {
            return changeStudentAdvisor(state, payload.advisor)
        }
        case CHANGE_CREDITS_NEEDED: {
            return changeStudentCreditsNeeded(state, payload.credits)
        }
        case CHANGE_MATRICULATION: {
            return changeStudentMatriculation(state, payload.matriculation)
        }
        case CHANGE_GRADUATION: {
            return changeStudentGraduation(state, payload.graduation)
        }
        case CHANGE_SETTING: {
            return changeStudentSetting(state, payload.key, payload.value)
        }

        case ADD_AREA: {
            return addAreaToStudent(state, payload.area)
        }
        case REMOVE_AREA: {
            return removeAreaFromStudent(state, payload.areaQuery)
        }
        case REMOVE_AREAS: {
            for (const areaQuery of payload.areaQueries) {
                state = removeAreaFromStudent(state, areaQuery)
            }
            return state
        }

        case ADD_SCHEDULE: {
            return addScheduleToStudent(state, payload.schedule)
        }
        case DESTROY_SCHEDULE: {
            return destroyScheduleFromStudent(state, payload.scheduleId)
        }
        case DESTROY_SCHEDULES: {
            for (const id of payload.scheduleIds) {
                state = destroyScheduleFromStudent(state, id)
            }
            return state
        }
        case RENAME_SCHEDULE: {
            return renameScheduleInStudent(
                state,
                payload.scheduleId,
                payload.newTitle
            )
        }
        case REORDER_SCHEDULE: {
            return reorderScheduleInStudent(
                state,
                payload.scheduleId,
                payload.newIndex
            )
        }
        case MOVE_SCHEDULE: {
            return moveScheduleInStudent(state, payload.scheduleId, {
                year: payload.year,
                semester: payload.semester,
            })
        }

        case ADD_COURSE: {
            return addCourseToSchedule(state, payload.scheduleId, payload.clbid)
        }
        case REMOVE_COURSE: {
            return removeCourseFromSchedule(
                state,
                payload.scheduleId,
                payload.clbid
            )
        }
        case REORDER_COURSE: {
            return reorderCourseInSchedule(state, payload.scheduleId, {
                clbid: payload.clbid,
                index: payload.index,
            })
        }
        case MOVE_COURSE: {
            return moveCourseToSchedule(state, {
                fromScheduleId: payload.fromScheduleId,
                toScheduleId: payload.toScheduleId,
                clbid: payload.clbid,
            })
        }

        case SET_OVERRIDE: {
            return setOverrideOnStudent(state, payload.key, payload.value)
        }
        case REMOVE_OVERRIDE: {
            return removeOverrideFromStudent(state, payload.override)
        }
        case ADD_FABRICATION: {
            return addFabricationToStudent(state, payload.fabrication)
        }
        case REMOVE_FABRICATION: {
            return removeFabricationFromStudent(state, payload.fabricationId)
        }

        default: {
            return state
        }
    }
}

export default undoable(studentReducer, {
    limit: 10,

    filter(action, currentState, previousState) {
        // only save history when something has changed.
        return currentState !== previousState
    },

    // treat LOAD_STUDENTS as the beginning of history
    initTypes: [
        '@@redux/INIT',
        '@@INIT',
        LOAD_STUDENT,
        INIT_STUDENT,
        IMPORT_STUDENT,
    ],
})
