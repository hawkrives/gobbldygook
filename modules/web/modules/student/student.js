import React, { Component, PropTypes, cloneElement } from 'react'
import DocumentTitle from 'react-document-title'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { loadStudent } from '../../redux/students/actions/load-student'

import Sidebar from '../../components/sidebar'
import Loading from '../../components/loading'

import CourseTable from '../course-table'
import GraduationStatus from './graduation-status-container'

import './student.scss'

export class Student extends Component {
    static propTypes = {
        content: PropTypes.node, // from react-router
        loadStudent: PropTypes.func.isRequired,
        overlay: PropTypes.node,
        params: PropTypes.object, // react-router
        processed: PropTypes.object, // redux
        sidebar: PropTypes.node, // from react-router
        student: PropTypes.object, // redux
    };

    componentWillMount() {
        this.loadStudent(this.props)
    }

    componentWillReceiveProps(nextProps) {
        this.loadStudent(nextProps)
    }

    loadStudent = props => {
        if (
            !props.student ||
            props.params.studentId !== this.props.params.studentId
        ) {
            props.loadStudent(props.params.studentId)
        }
    };

    render() {
        if (!this.props.student) {
            return (
                <div>
                    Student {this.props.params.studentId} could not be loaded.
                </div>
            )
        }

        if (this.props.student.isLoading) {
            return <Loading>Loading Student…</Loading>
        }

        const name = this.props.student
            ? this.props.student.data.present.name
            : 'Loading…'

        const contentProps = {
            student: this.props.student,
            className: 'content',
        }
        const contents = this.props.content
            ? cloneElement(this.props.content, contentProps)
            : <CourseTable {...contentProps} />

        const sidebarProps = { student: this.props.student.data.present }
        const sidebar = this.props.sidebar
            ? cloneElement(this.props.sidebar, sidebarProps)
            : <GraduationStatus {...sidebarProps} />

        return (
            <DocumentTitle title={`${name} | Gobbldygook`}>
                <div className="student">
                    <Sidebar student={this.props.student}>
                        {sidebar}
                    </Sidebar>
                    {contents}
                    {this.props.overlay || null}
                </div>
            </DocumentTitle>
        )
    }
}

const mapStateToProps = (state, ownProps) => ({
    student: state.students[ownProps.params.studentId],
})

const mapDispatchToProps = dispatch =>
    bindActionCreators({ loadStudent }, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(Student)
