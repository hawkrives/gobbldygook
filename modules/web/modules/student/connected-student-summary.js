// @flow

import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import {
    changeName,
    changeMatriculation,
    changeGraduation,
} from '../../redux/students/actions/change'

import { StudentSummary } from './student-summary'

type Student = Object

class WrappedStudentSummary extends React.PureComponent {
    props: {
        changeGraduation: string => any,
        changeMatriculation: string => any,
        changeName: string => any,
        student: Student,
    }

    handleChangeGraduation = (value: string) => {
        const newGraduation = parseInt(value) || 0
        this.props.changeGraduation(this.props.student.id, newGraduation)
    }

    handleChangeMatriculation = (value: string) => {
        const newMatriculation = parseInt(value) || 0
        this.props.changeMatriculation(this.props.student.id, newMatriculation)
    }

    handleChangeName = (value: string) => {
        this.props.changeName(this.props.student.id, value)
    }

    render() {
        return (
            <StudentSummary
                onChangeGraduation={this.handleChangeGraduation}
                onChangeMatriculation={this.handleChangeMatriculation}
                onChangeName={this.handleChangeName}
                student={this.props.student}
            />
        )
    }
}

const mapDispatch = dispatch =>
    bindActionCreators(
        {
            changeName,
            changeMatriculation,
            changeGraduation,
        },
        dispatch
    )

// $FlowFixMe
export const ConnectedStudentSummary = connect(null, mapDispatch)(
    WrappedStudentSummary
)
