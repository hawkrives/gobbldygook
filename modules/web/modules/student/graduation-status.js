// @flow
import React from 'react'

import { AreaOfStudySidebar } from './area-of-study-sidebar'
import { ConnectedStudentSummary } from './connected-student-summary'

import './graduation-status.scss'

type Student = Object

export default class GraduationStatusContainer extends React.PureComponent {
    props: {
        student: Student,
    }

    render() {
        const student = this.props.student
        if (!student) {
            return null
        }

        return (
            <section className="graduation-status">
                <ConnectedStudentSummary student={student} />
                <AreaOfStudySidebar student={student} />
            </section>
        )
    }
}
