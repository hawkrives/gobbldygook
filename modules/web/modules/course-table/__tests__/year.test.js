import React from 'react'
import Year from '../year'
import { Student } from '../../../../../modules/object-student/student'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

const mockStudent = () =>
    Student({
        name: 'test',
        id: '0xabadidea',
        matriculation: 2015,
        graduation: 2019,
        dateCreated: new Date('2017-03-17T02:03:33.974Z'),
        dateLastModified: new Date('2017-03-17T02:03:33.974Z'),
    })

test('Year renders shallowly', () => {
    const addSemester = jest.fn()
    const removeYear = jest.fn()

    const tree = renderer.create(
        <Year
            addSemester={addSemester}
            removeYear={removeYear}
            student={mockStudent()}
            year={2016}
        />
    )

    expect(tree).toMatchSnapshot()
})

test('Year can add a semester', () => {
    const addSemester = jest.fn()
    const removeYear = jest.fn()

    const student = mockStudent()

    const tree = shallow(
        <Year
            addSemester={addSemester}
            removeYear={removeYear}
            student={student}
            year={2016}
        />
    )

    expect(tree).toMatchSnapshot()

    tree.find({ title: 'Add Semester' }).simulate('click')

    expect(addSemester).toHaveBeenCalled()
    expect(removeYear).not.toHaveBeenCalled()

    expect(tree).toMatchSnapshot()
})

test('Year can remove a semester', () => {
    const addSemester = jest.fn()
    const removeYear = jest.fn()

    const student = mockStudent()

    const tree = shallow(
        <Year
            addSemester={addSemester}
            removeYear={removeYear}
            student={student}
            year={2016}
        />
    )

    expect(tree).toMatchSnapshot()

    tree.find({ title: 'Remove the year 2016â€”2017' }).simulate('click')

    expect(addSemester).not.toHaveBeenCalled()
    expect(removeYear).toHaveBeenCalled()

    expect(tree).toMatchSnapshot()
})
