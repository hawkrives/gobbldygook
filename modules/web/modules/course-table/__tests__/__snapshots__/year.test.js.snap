// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Year can add a semester 1`] = `
<styled.div>
  <styled.header>
    <styled.h1>
      2016—2017
    </styled.h1>
    <span>
      <Styled(Button)
        onClick={[Function]}
        title="Add Semester"
        type="flat"
      >
        Add ‘
        Fall
        ’
      </Styled(Button)>
      <Styled(Styled(Button))
        onClick={[Function]}
        title="Remove the year 2016—2017"
        type="flat"
      >
        Remove Year
      </Styled(Styled(Button))>
    </span>
  </styled.header>
  <styled.div />
</styled.div>
`;

exports[`Year can add a semester 2`] = `
<styled.div>
  <styled.header>
    <styled.h1>
      2016—2017
    </styled.h1>
    <span>
      <Styled(Button)
        onClick={[Function]}
        title="Add Semester"
        type="flat"
      >
        Add ‘
        Fall
        ’
      </Styled(Button)>
      <Styled(Styled(Button))
        onClick={[Function]}
        title="Remove the year 2016—2017"
        type="flat"
      >
        Remove Year
      </Styled(Styled(Button))>
    </span>
  </styled.header>
  <styled.div />
</styled.div>
`;

exports[`Year can remove a semester 1`] = `
<styled.div>
  <styled.header>
    <styled.h1>
      2016—2017
    </styled.h1>
    <span>
      <Styled(Button)
        onClick={[Function]}
        title="Add Semester"
        type="flat"
      >
        Add ‘
        Fall
        ’
      </Styled(Button)>
      <Styled(Styled(Button))
        onClick={[Function]}
        title="Remove the year 2016—2017"
        type="flat"
      >
        Remove Year
      </Styled(Styled(Button))>
    </span>
  </styled.header>
  <styled.div />
</styled.div>
`;

exports[`Year can remove a semester 2`] = `
<styled.div>
  <styled.header>
    <styled.h1>
      2016—2017
    </styled.h1>
    <span>
      <Styled(Button)
        onClick={[Function]}
        title="Add Semester"
        type="flat"
      >
        Add ‘
        Fall
        ’
      </Styled(Button)>
      <Styled(Styled(Button))
        onClick={[Function]}
        title="Remove the year 2016—2017"
        type="flat"
      >
        Remove Year
      </Styled(Styled(Button))>
    </span>
  </styled.header>
  <styled.div />
</styled.div>
`;

exports[`Year renders shallowly 1`] = `
<div
  className="sc-jhAzac jHxpTp"
>
  <header
    className="sc-fBuWsC hJifQM"
  >
    <h1
      className="sc-fMiknA lvJZL"
    >
      2016—2017
    </h1>
    <span>
      <button
        className="button button--flat sc-dVhcbM fvKlhn"
        disabled={undefined}
        onClick={[Function]}
        style={undefined}
        title="Add Semester"
        type="button"
      >
        Add ‘
        Fall
        ’
      </button>
      <button
        className="button button--flat sc-eqIVtm gjViaw sc-dVhcbM fvKlhn"
        disabled={undefined}
        onClick={[Function]}
        style={undefined}
        title="Remove the year 2016—2017"
        type="button"
      >
        Remove Year
      </button>
    </span>
  </header>
  <div
    className="sc-fAjcbJ kuQnpu"
  />
</div>
`;
